File: Medium/207. Course Schedule/solution.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        vector<int> inDegree(numCourses, 0);
        
        for (int i = 0; i < prerequisites.size(); i++){
            graph[prerequisites[i][1]].push_back(prerequisites[i][0]);
            inDegree[prerequisites[i][0]]++;
        }
        
        queue<int> myQueue;
        for (int i = 0; i < numCourses; i++){
            if (inDegree[i] == 0){
                myQueue.push(i);
            }
        }
        
        while (!myQueue.empty()){
            int curCourse = myQueue.front();
            myQueue.pop();
            
            
            for (int nextCourse: graph[curCourse]){
                inDegree[nextCourse]--;
                
                if (inDegree[nextCourse] == 0){
                    myQueue.push(nextCourse);
                }
            }
        }
        
        for (int i = 0; i < numCourses; i++){
            if (inDegree[i] != 0)   return false;
        }
        return true;
    }
};

File: Medium/207. Course Schedule/solution.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        inDegree = [0] * numCourses
        graph = [[] for _ in range(numCourses)]

        for i in range(len(prerequisites)):
            inDegree[prerequisites[i][0]] += 1
            graph[prerequisites[i][1]].append(prerequisites[i][0])

        q = deque()
        for i in range(numCourses):
            if inDegree[i] == 0:
                q.append(i)
        
        n = 0
        while q:
            cur = q.pop()
            n += 1
            for i in range(len(graph[cur])):
                inDegree[graph[cur][i]] -= 1
                if inDegree[graph[cur][i]] == 0:
                    q.append(graph[cur][i])

        return n == numCourses
        